### PR TITLE
Feat/save as

### PR DESCRIPTION
**Pull request by tonsV2**
Please see the below jira issue for details.
https://jira.dhis2.org/browse/DEVOPS-175

A good starting point for a reviewer would perhaps be the user script at scripts/saveAs.sh

# TODO

*         - name: data
          emptyDir:
            sizeLimit: {{ .Values.dataSizeLimit }}
dataSizeLimit seems to have no effect and the mounted disk doesn't seem to be a volume... But it is writable

* Timeout... Eventually we'll need to run the export etc. in a background thread. Should we just return 202 or an id or something else?
* Don't implement the endpoint on the database manager. Rather implement /instances/:id/save-as on the manager and let that produce an event which will be consumed by the database manager
* Add support for remote clusters
* Don't corrupt gz file!!!

### CODE REVIEW

**Comment by Philip-Larsen-Donnelly** [comment]
Timeout: shouldn't we ideally be returning the id of the job (with a 202) so that we can check the progress. The same will be true for any action/event we want to run against an instance.

**Comment by tonsV2** [comment]
> Timeout: shouldn't we ideally be returning the id of the job (with a 202) so that we can check the progress. The same will be true for any action/event we want to run against an instance.

I'm not running this as a job. The jobs implementation isn't completed yet 

How about doing something like the following. At the beginning of the request I'll save the database for the sake of populating the id column which can then be used to track the state, I also introduce a status column which I'll update as the request progresses. I'll run the time consuming operations in the background and return the id along with a 202?

The above description is a bit simplified but I think it explains the overall idea.

File: pkg/database/service.go [hunk 1]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+
+	dump, err := newPgDumpConfig(instance, stack)
+	if err != nil {
+		return nil, err
+	}
+
+	var ret *forwarder.Result
+	if len(group.ClusterConfiguration.KubernetesConfiguration) > 0 {
+		hostname := fmt.Sprintf(stack.HostnamePattern, instance.Name, instance.GroupName)
+		serviceName := strings.Split(hostname, ".")[0]
```
Comment (teleivo) [inline pkg/database/service.go:256]: Do I get this right that for example `hostnamePattern: %s-postgresql.%s.svc` the serviceName is `%s-postgresql` => `${instance.Name}-postgresql`
excuse my pseudocode 😂

File: pkg/database/service.go [hunk 2]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+		}
+
+		err = os.Remove(dumpFile)
+		if err != nil {
+			//			return nil, err
+			log.Println(err)
+			return
+		}
+
+		ret.Close()
```
Comment (teleivo) [inline pkg/database/service.go:359]: I think we might be leaking ports on error as we don't close using `defer` here. What statement is actually starting the port forwarding?

File: pkg/database/service.go [hunk 3]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+	}
+
+	err = s.repository.Save(d)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		// TODO: Remove... Or at least make configurable
+		dump.EnableVerbose()
```
Comment (teleivo) [inline pkg/database/service.go:302]: Can this potentially print sensitive information? Not sure what the dump emits during its process 😄

File: pkg/database/service.go [hunk 4]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+
+		dumpFile := dumpPath + dumpExec.File
+		file, err := os.Open(dumpFile)
+		if err != nil {
+			//			return nil, err
+			log.Println(err)
+			return
+		}
+
+		if format == "plain" {
```
Comment (teleivo) [inline pkg/database/service.go:321]: Can you add a validation in the controller for the formats we support? You have validation for some params that need to be valid DNS names I think

File: pkg/database/service.go [hunk 5]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+		}
+
+		_, err = s.Upload(d, group, file)
+		if err != nil {
+			//			return nil, err
+			log.Println(err)
+			return
+		}
+
+		err = file.Close()
```
Comment (teleivo) [inline pkg/database/service.go:345]: If we don't close it in a defer we need to close it in every (error) case 😬 
For example if `s.Upload(d, group, file)` we won't close the file.

File: pkg/database/service.go [hunk 6]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+			return
+		}
+
+		ret.Close()
+	}()
+
+	return d, nil
+}
+
+func gz(gzFile string, database *model.Database, file *os.File) (*os.File, error) {
```
Comment (teleivo) [inline pkg/database/service.go:365]: Maybe call `file` => `src` or so to make it more clear 😄 We different files floating around only using `file` also in the method above makes it a bit hard to follow.

File: pkg/database/service.go [hunk 7]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+			gzFile := dumpPath + fileId + ".gz"
+			file, err = gz(gzFile, database, file)
+			if err != nil {
+				//			return nil, err
+				log.Println(err)
+				return
+			}
+
+			defer func() {
+				err = os.Remove(gzFile)
```
Comment (teleivo) [inline pkg/database/service.go:331]: I think the file needs to be close first, otherwise it might not be removed. See https://stackoverflow.com/questions/64943210/should-you-close-temporarily-created-files-or-can-you-directly-delete-them (its about temp files originally, but the answer goes into how an OS treats files that are still referenced). We had this some time back with regards to the kubeconfig if you remember 😋

File: pkg/database/service.go [hunk 8]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+	if err != nil {
+		return nil, err
+	}
+
+	err = file.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	file, err = os.Open(gzFile)
```
Comment (teleivo) [inline pkg/database/service.go:389]: Why are we opening the file here? And overwriting the `file`?
I would not expect a function that gzips a file to leave it open when it returns.

File: pkg/database/service.go [hunk 9]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+
+	zw := gzip.NewWriter(outFile)
+	zw.Name = strings.TrimSuffix(database.Name, ".gz")
+
+	_, err = io.Copy(zw, file)
+	if err != nil {
+		return nil, err
+	}
+
+	err = zw.Close()
```
Comment (teleivo) [inline pkg/database/service.go:379]: As above if closing `zw` fails we will not close `file` since we are not using defer.

File: pkg/database/service.go [hunk 10]
```diff
@@ -224,3 +235,210 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+		// TODO: For now, only saving to the same group is supported
+		GroupName: instance.GroupName,
+	}
+
+	err = s.repository.Save(d)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
```
Comment (teleivo) [inline pkg/database/service.go:300]: pgdump could keep the goroutine up for a long time, maybe even never free it if it can get stuck. We also have no limit on how many of these we start. I assume we are ok with these risks for now, right?

**Review by teleivo — COMMENTED** [review]
There are lots of files in play 😅 I am not 100% sure we are closing/removing them reliably.

As the code keeps on growing, I really think you should add some tests 😬 

An endpoint for getting a status on the dump is going to come later, right?

File: pkg/database/service.go (hunk 1 above)
Comment (tonsV2) [reply pkg/database/service.go:256]: Yes, the hostname convention in Kubernetes is service-name.namespace.svc and here I'm extracting the service name

File: pkg/database/service.go (hunk 3 above)
Comment (tonsV2) [reply pkg/database/service.go:302]: I'm not going to bet my life on it, but I don't think so. As the comment indicates I intent to deal with this differently but due to the current lack of insights I hard coded it for now :-)

File: pkg/database/service.go (hunk 4 above)
Comment (tonsV2) [reply pkg/database/service.go:321]: I think we have to implement a custom validator but it shouldn't be that hard

File: pkg/database/service.go (hunk 10 above)
Comment (tonsV2) [reply pkg/database/service.go:300]: Yes, we should, and will, handle this differently but for now we'll accept the risks. This "fire and forget" approach is also how it's done on the AWX based system.

File: README.md [hunk 11]
```diff
@@ -14,3 +13,16 @@ wget https://databases.dhis2.org/sierra-leone/2.36.0/dhis2-db-sierra-leone.sql.g
 ```sh
 export ACCESS_TOKEN=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjQ3OTYzODQwNzcsImlhdCI6MTY0MDY5MTQ3NywidXNlciI6eyJJRCI6NCwiQ3JlYXRlZEF0IjoiMjAyMS0xMi0yOFQxMDo0NjoxNi44NTEzMzlaIiwiVXBkYXRlZEF0IjoiMjAyMS0xMi0yOFQxMDo0NjoxNi44NTEzMzlaIiwiRGVsZXRlZEF0IjpudWxsLCJFbWFpbCI6InNvbWVvbmVAc29tZXRoaW5nLm9yZyIsIkdyb3VwcyI6bnVsbCwiQWRtaW5Hcm91cHMiOm51bGx9fQ.FnPIu36kV1T-Jix5Wy-HsZeqxQI6Q_7HQ14C1DWKHETIBSk-vLQ_sCMHVPKA42utEDFI3Xpmf6Gyzv9aPU_Cvg-JDazRprfrZBqn4LzSmT6K3HGoKoQ0b5G8exxz0Ote8NQDB1NBZmYvD1gpVVisCvzaewJTRAvRA3DS0n_O4kU5QENdLNPfWFo0rXOC83sLBsEIe2Ce4TiRrepOCSQKE-_rwQQSA3w30MhFmhAY7Ozcd9i69mtfcvqjORdNJ-zREgiw8B2g9oh7byE1h2oxjvoKC3WRfPeSYoRY6GuMHSSWJdzFKIswlZHdWU1GicPJASBbkKGbP5n5O6FXyeo0bw
 ```
+
+
+# TODO
+
+*         - name: data
```
[the remark this answers is not in the document]
Comment (tonsV2) [reply README.md:20]: Yes

File: helm/chart/values.yaml [hunk 12]
```diff
@@ -32,6 +35,10 @@ userService:
   host: user-service.namespace.svc
   basePath:
 
+instanceService:
+  host: manager-service.namespace.svc
+  basePath:
```
[the remark this answers is not in the document]
Comment (tonsV2) [reply helm/chart/values.yaml:40]: It's deliberate :-)

File: pkg/database/service.go (hunk 2 above)
Comment (tonsV2) [reply pkg/database/service.go:359]: The port forwarding is instantiated [here](https://github.com/dhis2-sre/im-database-manager/blob/df70ec87af4c2ce450e7fafb156a97d7ff0d4bc9/pkg/database/service.go#L270)  and we wait for it to become ready [here](https://github.com/dhis2-sre/im-database-manager/blob/df70ec87af4c2ce450e7fafb156a97d7ff0d4bc9/pkg/database/service.go#L275) if that matters.

My reason for not using defer in the first place is that we don't want to close the port before the go routine is done.

File: pkg/database/service.go (hunk 8 above)
Comment (tonsV2) [reply pkg/database/service.go:389]: If the format is `plain` and we want to gzip the output file. Then the output file is the new gzipped file and not the previous one which I close a few lines up.

The reason for returning an open file in the first place is the upload method taking a `io.Reader` as input. I'm not sure if that's something we want to change.

File: pkg/database/service.go (hunk 1 above)
Comment (teleivo) [reply pkg/database/service.go:256]: If this is something we do regularly we should abstract this away somewhere 😄

File: pkg/database/service.go (hunk 2 above)
Comment (teleivo) [reply pkg/database/service.go:359]: Is the port forwarding started when instantiating or when we call `ret.Ready()`? This matters with regards to from which point on you need to make sure to close port forwarding.

> My reason for not using defer in the first place is that we don't want to close the port before the go routine is done.

You could use defer in the function running the go routine though. If you decide to not use defer you need to think about what it does for you and then do that yourself 😋 This means you will need to call `ret.Close()` or whatever you defer in all cases. For example when `s.Upload(d, group, file)`, `file.Close()`, `os.Remove(dumpFile)`, ... fails

File: pkg/database/service.go (hunk 4 above)
Comment (teleivo) [reply pkg/database/service.go:321]: Whats the validator we are using? There is no "these few values are the only valid ones" out of the box 😬

File: pkg/database/service.go (hunk 8 above)
Comment (teleivo) [reply pkg/database/service.go:389]: > If the format is plain and we want to gzip the output file. Then the output file is the new gzipped file and not the previous one which I close a few lines up.

I think you could return `outFile` instead? Which would make it easier to follow.

> The reason for returning an open file in the first place is the upload method taking a io.Reader as input. I'm not sure if that's something we want to change.

I would just not see this responsibility inside of `gz`.

File: cmd/serve/main.go [hunk 13]
```diff
@@ -76,6 +79,28 @@ func run() error {
...
+	if err != nil {
+		return err
+	}
+
 	r := server.GetEngine(cfg.BasePath, dbHandler, authMiddleware)
 	return r.Run()
 }
+
+// Inspiration: https://blog.logrocket.com/gin-binding-in-go-a-tutorial-with-examples/
+func registerValidation() error {
```
Comment (teleivo) [inline cmd/serve/main.go]: You agreed to move this out of main 😋 into internal handler package I think.

File: internal/handler/validation_test.go [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+	err = ctx.ShouldBind(&TestRequest{OneOf: "one"})
+	assert.NoError(t, err)
+
+	err = ctx.ShouldBind(&TestRequest{OneOf: "of"})
+	assert.NoError(t, err)
+
+	err = ctx.ShouldBind(&TestRequest{OneOf: "oh no"})
+	assert.Error(t, err)
+	assert.Equal(t, "Key: 'TestRequest.OneOf' Error:Field validation for 'OneOf' failed on the 'oneOf' tag", err.Error())
```
Comment (teleivo) [inline internal/handler/validation_test.go:34]: Do we have control over this message? Would you as a user know what to do? I wouldn't 😬 I would expect such an error message to say something like `got "oh no" in field `xxx`, only one of "one", "of" is allowed.`

File: internal/handler/validation.go [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+	"fmt"
+	"github.com/gin-gonic/gin/binding"
+	"github.com/go-playground/validator/v10"
+	"strings"
+)
+
+// RegisterValidation Inspiration: https://blog.logrocket.com/gin-binding-in-go-a-tutorial-with-examples/
+func RegisterValidation() error {
+	if v, ok := binding.Validator.Engine().(*validator.Validate); ok {
+		return v.RegisterValidation("oneOf", func(fl validator.FieldLevel) bool {
```
Comment (teleivo) [inline internal/handler/validation.go:13]: I would create a named function `oneOf`. The `RegisterValidation` then just registers potentially multiple functions. Also has the benefit of being able to unit test `oneOf` if there are many cases that we do not want to all cover via an httptest test.

File: internal/handler/validation_test.go [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+	"testing"
+)
+
+type TestRequest struct {
+	OneOf string `binding:"required,oneOf=one of"`
+}
+
+func TestRegisterValidation(t *testing.T) {
+	err := RegisterValidation()
+	assert.NoError(t, err)
```
Comment (teleivo) [inline internal/handler/validation_test.go:17]: I think we should use require here as if registration fails there is no point in testing anything else as it won't work anyway.

File: internal/handler/validation_test.go [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+type TestRequest struct {
+	OneOf string `binding:"required,oneOf=one of"`
```
Comment (teleivo) [inline internal/handler/validation_test.go:12]: I would not reuse the name `oneOf` for anything like struct field and values. This was confusing to me when I first read the test and the error message. Maybe just call the field `field` and the values `binding:"required,oneOf=one two"`.

File: internal/handler/validation_test.go [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+func TestRegisterValidation(t *testing.T) {
+	err := RegisterValidation()
+	assert.NoError(t, err)
+
+	w := httptest.NewRecorder()
+	ctx, _ := gin.CreateTestContext(w)
+
+	dummyRequest, err := http.NewRequest("GET", "/", nil)
```
Comment (teleivo) [inline internal/handler/validation_test.go:22]: I try not to prefix things as mock, dummy, test or so in tests. I think its clear that we are in a test and some things might not be 100% of what they would be in production code.

File: internal/handler/validation_test.go [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+type TestRequest struct {
```
Comment (teleivo) [inline internal/handler/validation_test.go:11]: does this really represent a request?

File: pkg/database/service.go [hunk 20]
```diff
@@ -224,3 +235,202 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+			logError(err)
+			return
+		}
+	}()
+
+	return d, nil
+}
+
+func logError(err error) {
+	// TODO: Persist error message
```
Comment (teleivo) [inline pkg/database/service.go:346]: what do you mean by that?

File: pkg/database/service.go [hunk 21]
```diff
@@ -224,3 +235,202 @@ func (s service) FindExternalDownload(uuid uuid.UUID) (model.ExternalDownload, e
...
+	if err != nil {
+		return nil, err
+	}
+
+	dump, err := newPgDumpConfig(instance, stack)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &model.Database{
```
Comment (teleivo) [inline pkg/database/service.go:253]: Either we declare this close to where we return it or we give it a longer name maybe `database`. Looking at the last return in this long method I forgot what `d` is.

**Review by teleivo — COMMENTED** [review]
Getting close 😄

File: cmd/serve/main.go (hunk 13 above)
Comment (teleivo) [reply cmd/serve/main.go]: GitHub did not show me that you actually did at the beginning 🤷🏻 so this is done 😄

File: internal/handler/validation_test.go (hunk 19 above)
Comment (tonsV2) [reply internal/handler/validation_test.go:11]: Would you find payload a more suitable name?

File: pkg/database/service.go (hunk 20 above)
Comment (tonsV2) [reply pkg/database/service.go:346]: Currently we just print the error to stdout. In the future we quite likely want to persist the error on the database model

File: internal/handler/validation_test.go (hunk 16 above)
Comment (tonsV2) [reply internal/handler/validation_test.go:17]: What do you mean by require? The test fails if there's an error

Comment (teleivo) [reply internal/handler/validation_test.go:17]: Yes the test will fail if there is an error but the code below this line will still be executed. We don't need to execute the code below as it very likely cannot succeed.

testify has assertions in packages `assert` and `require`. The former is like calling `t.Error` (fail the test but execute anything that comes after it) while the latter is like calling `t.Fatal`/`t.Fail` (fail the test right away).

**Review by teleivo — APPROVED** [review]
